src/Tmds.Ssh/SftpClient.cs

```diff
@@ -0,0 +1,214 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Concurrent;
+using System.Buffers;
+using System.Buffers.Binary;
+
+namespace Tmds.Ssh
+{
+    abstract class SftpOperation
+    {
+        public abstract ValueTask HandleResponse(SftpPacketType type, ReadOnlySequence<byte> fields);
+
+        protected static Exception CreateExceptionForStatus(ReadOnlySequence<byte> fields)
+        {
+            return new SshException("TODO"); // TODO: use fields
+        }
+
+        protected static Exception CreateExceptionForUnexpectedType(SftpPacketType type)
+        {
+            return new ProtocolException($"Unexpected response type: {type}");
+        }
+    }
+
+    public sealed partial class SftpClient : IDisposable
+    {
+        // SSH_MSG_CHANNEL_DATA          (1)
+        // RemoteChannel                 (4)
+        // SSH_MSG_CHANNEL_DATA length   (4)
+        // SFTP length                   (4)
+        const int DataHeaderLength = 13;
+        const int SftpVersion = 3;
+        private readonly ChannelContext _context;
+        private Task? _receiveLoopTask;
+        private int _requestId;
+        private readonly ConcurrentDictionary<uint, SftpOperation> _operations;
+
+        private ValueTask SendRequestAsync(Packet packet, SftpOperation operation)
+        {
+            /*
+                0: SSH_MSG_CHANNEL_DATA  (1)
+                1: RemoteChannel         (4)
+                5: DATA length           (4)
+                9: SFTP length           (4)
+                13: SFTP packet type     (1) // Already filled in.
+                14: SFTP request id      (4)
+            */
+            uint requestId = unchecked((uint)Interlocked.Increment(ref _requestId));
+            Span<byte> header = packet.PayloadHeader;
+            header[0] = (byte)MessageId.SSH_MSG_CHANNEL_DATA;
+            BinaryPrimitives.WriteUInt32BigEndian(header.Slice(1), _context.RemoteChannel);
+            int dataLength = (int)packet.PayloadLength - 9;
+            BinaryPrimitives.WriteUInt32BigEndian(header.Slice(5), (uint)dataLength);
+            BinaryPrimitives.WriteUInt32BigEndian(header.Slice(9), (uint)(dataLength - 4));
+            BinaryPrimitives.WriteUInt32BigEndian(header.Slice(14), requestId);
+
+            _operations.TryAdd(requestId, operation);
+
+            return _context.SendChannelDataAsync(packet, default(CancellationToken)); // TODO: cancellation
+        }
+
+        internal SftpClient(ChannelContext context)
+        {
+            _context = context;
+            _operations = new ConcurrentDictionary<uint, SftpOperation>();
+        }
+
+        public async Task InitAsync(CancellationToken ct)
+        {
+            await SftpInitMessageAsync(3, ct).ConfigureAwait(false);
+            // TODO add server negotiation in case server would have min. version < 3, but was able to do a version 3 aswell
```

Comment:
Is there a TODO here? It's ok if we only work with servers that support version 3 and above.
Since we've sent 3, those servers should reply with 3 also:

>    'version' is the lower of the protocol version supported by the
   server and the version number received from the client.

Reply:
Yes I've misunderstood the negotiation, for some reason I thought they both will supply its on lowest spec support when I've read it first time. But that doesn't make sense, will remove